File: js/Balanced_Binary_Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */

function getDepth(root) {
    if (root === null) {
        return 0;
    }
    return 1 + Math.max(getDepth(root.left), getDepth(root.right));
}

var isBalanced = function(root) {
    if (root === null) {
        return true;
    }
    if (Math.abs(getDepth(root.left) - getDepth(root.right)) > 1) {
        return false;
    }
    // 所有子树都必须是平衡的
    return isBalanced(root.left) && isBalanced(root.right);
};

File: go/Balanced_Binary_Tree.go
package main

import "math"

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func getDepth(node *TreeNode) int {
    if node == nil {
        return 0
    }
    depth := 1 + math.Max(float64(getDepth(node.Left)), float64(getDepth(node.Right)))
    return int(depth)
}

func isBalanced(root *TreeNode) bool {
    if root == nil {
        return true
    }
    if math.Abs(float64(getDepth(root.Left)) - float64(getDepth(root.Right))) > 1 {
        return false
    }
    return isBalanced(root.Left) && isBalanced(root.Right)
}
